lib/Sema/DerivedConformances.cpp

```diff
@@ -380,3 +389,215 @@ bool DerivedConformance::checkAndDiagnoseDisallowedContext(
+    // generate: indexVar = <index>
+    auto indexExpr = buildIntegerLiteral(C, index++);
+
+    auto indexRef = new (C) DeclRefExpr(indexVar, DeclNameLoc(),
+                                        /*implicit*/true,
+                                        AccessSemantics::Ordinary,
+                                        LValueType::get(intType));
+    auto assignExpr = new (C) AssignExpr(indexRef, SourceLoc(),
+                                         indexExpr, /*implicit*/ true);
+    assignExpr->setType(TupleType::getEmpty(C));
+    auto body = BraceStmt::create(C, SourceLoc(), ASTNode(assignExpr),
+                                  SourceLoc());
+    cases.push_back(CaseStmt::create(C, SourceLoc(), labelItem, SourceLoc(),
+                                     SourceLoc(), body,
+                                     /*case body vardecls*/ None));
+  }
+
+  // generate: switch enumVar { }
+  auto enumRef = new (C) DeclRefExpr(enumVarDecl, DeclNameLoc(),
+                                     /*implicit*/true,
+                                     AccessSemantics::Ordinary,
+                                     enumVarDecl->getType());
+  auto switchStmt = SwitchStmt::create(LabeledStmtInfo(), SourceLoc(), enumRef,
+                                       SourceLoc(), cases, SourceLoc(), C);
+
+  stmts.push_back(indexBind);
+  stmts.push_back(switchStmt);
+
+  return new (C) DeclRefExpr(indexVar, DeclNameLoc(), /*implicit*/ true,
+                             AccessSemantics::Ordinary, intType);
+}
+
+/// Returns the ParamDecl for each associated value of the given enum whose type
+/// does not conform to a protocol
+/// \p theEnum The enum whose elements and associated values should be checked.
+/// \p protocol The protocol being requested.
+/// \return The ParamDecl of each associated value whose type does not conform.
+SmallVector<ParamDecl *, 3>
+swift::associatedValuesNotConformingToProtocol(DeclContext *DC, EnumDecl *theEnum,
+                                        ProtocolDecl *protocol) {
+  SmallVector<ParamDecl *, 3> nonconformingAssociatedValues;
+  for (auto elt : theEnum->getAllElements()) {
+    auto PL = elt->getParameterList();
+    if (!PL)
+      continue;
+
+    for (auto param : *PL) {
+      auto type = param->getInterfaceType();
+      if (TypeChecker::conformsToProtocol(DC->mapTypeIntoContext(type),
+                                          protocol, DC, None)
+              .isInvalid()) {
+        nonconformingAssociatedValues.push_back(param);
+      }
+    }
+  }
+  return nonconformingAssociatedValues;
+}
+
+/// Returns true if, for every element of the given enum, it either has no
+/// associated values or all of them conform to a protocol.
+/// \p theEnum The enum whose elements and associated values should be checked.
+/// \p protocol The protocol being requested.
+/// \return True if all associated values of all elements of the enum conform.
+bool swift::allAssociatedValuesConformToProtocol(DeclContext *DC,
```

Comment:
Can we limit the scope of this API promotion to `DerivedConformance`?

Reply:
done